xcdat/dataset.py

```diff
@@ -643,3 +677,45 @@ def _split_time_units_attr(units_attr: str) -> Tuple[str, str]:
         )
 
     return units, reference_date
+
+
+def _get_cftime_coords(
+    ref_date: str, offsets: np.ndarray, calendar: str, units: str
+) -> np.ndarray:
+    """Get an array of `cftime` coordinates starting from a reference date.
+
+    Parameters
+    ----------
+    ref_date : str
+        The starting reference date.
+    offsets : np.ndarray
+        An array of numerically encoded time offsets from the reference date.
+    calendar : str
+        The CF calendar type supported by ``cftime``. This includes "noleap",
+        "360_day", "365_day", "366_day", "gregorian", "proleptic_gregorian",
+        "julian", "all_leap", and "standard".
+    units : str
+        The time units.
+
+    Returns
+    -------
+    np.ndarray
+        An array of `cftime` coordinates.
+    """
+    # Starting from the reference date, create an array of `datetime` objects
+    # by adding each offset (a numerically encoded value) to the reference date.
+    # The `parse.parse` default is set to datetime(2000, 1, 1), with each
+    # component being a placeholder if the value does not exist. For example, 1
+    # and 1 are placeholders for month and day if those values don't exist.
+    ref_datetime: datetime = parser.parse(ref_date, default=datetime(2000, 1, 1))
+    offsets = np.array(
+        [ref_datetime + rd.relativedelta(**{units: offset}) for offset in offsets],
+        dtype="datetime64",
```

Comment:
```suggestion
        dtype="object",
```

In the `convert_times()` method, there is a conditional that checks if the `offsets` np.ndarray dtype `is_np_datetime_like` (which then converts to pd.Timestamp which we don't want).

https://github.com/pydata/xarray/blob/3c98ec7d96cc4b46664850cc7a40af2bc184fea0/xarray/coding/times.py#L454-L460

Instead, by setting the `dtype=object`, it returns an np.ndarray of `cftime` objects.

https://github.com/pydata/xarray/blob/3c98ec7d96cc4b46664850cc7a40af2bc184fea0/xarray/coding/times.py#L461-L476